doc/hem.1.txt
HEM(1)
======
Ryan Tomayko <r@tomayko.com>

NAME
----
hem - Persistent SSH connection and tunnel manager.

SYNOPSIS
--------
[verse]
`hem` [`-qv`] [`-c` 'config_file'] 'command' ['command_options'] ['profile'...]
`hem` `init` [`-ef`] [`-d` 'dir']
`hem` `manage` [`-Ae`] 'profile'...
`hem` `status` [`-cp`] 'profile'...
`hem` `up` [`-fp`] 'profile'...
`hem` `down` [`-s` 'sig'] 'profile'...
`hem` `bounce` 'profile'...
`hem` `info` ['options'] 'profile'...

DESCRIPTION
-----------

`Hem` manages multiple background SSH connections using an man:ifconfig[8] /
man:rc[8] style interface. The `hem` program is a front-end to multiple
sub-commands: `init`, `manage`, `status`, `up`, `down`, `bounce`, and `info`.
See the *HEM COMMANDS* section below for a description of each command.

When `hem` is invoked, it reads configuration information from *~/.hem/config*,
or the file specified by the `--config` argument, or the value of the
`HEM_CONFIG` environment variable. See manlocal:hem_config[5] for information on
configuration values.

OPTIONS
-------
-c, --config 'config_file'::
	Load configuration from 'config_file' instead of the default configuration
	file location. The configuration file specifies the location of hem
	profiles and various other options that control its operation.
	See manlocal:hem_config[5] for more information on configuration options and
	format.

-q, --quiet::
	Do not write any non-critical output to *stdout* / *stderr*. Without the
	`--quiet` option, `hem` will write a small bit of informational
	output to *stderr* on what its doing.

-v, --verbose::
	Write verbose messages typical reserved for the log to *stderr* in
	addition to normal informational output.

--help::
	Show program usage and exit.

HEM COMMANDS
------------
Each hem sub-command is fully documented under its own manual page:

manlocal:hem-init[1]::
	Initialize a new configuration directory (`~/.hem` by default) and
	check that ssh configuration is amenable to use with Hem.

manlocal:hem-manage[1]::
	Add a new connection profile and optionally authorize with the
	remote host. This command can also be used to edit existing profiles.

manlocal:hem-info[1]::
	Show configuration information for one or more profiles.

manlocal:hem-status[1]::
	Show connection status for one or more profiles.

manlocal:hem-up[1]::
	Bring connections up for one or more profiles.

manlocal:hem-down[1]::
	Take connections down for one or more profiles.

manlocal:hem-bounce[1]::
	Restart or reload the connection for one or more profiles.

CONNECTION PROFILES
-------------------

All hem commands -- with the exception of manlocal:hem-init[1] -- operate on
one or more connection 'profiles'. Each 'profile' is represented by a file in
the profiles directory: *~/.hem/profile* (this can be overridden by setting
the `profile_dir` variable in manlocal:hem_config[5]).

A connection profile includes configuration for a single remote SSH
connection. The profile filename is used to provide a minimal amount of
connection information and is named similar to an SSH remote specifier:
['user'@]{empty}'hostname'[:'port']. For example, you may have profiles named
*example.com*, *root@example.com*, and *toor@example.com:522*.

EXAMPLES
--------
Use the manlocal:hem-init[1] command to initialize a new configuration
directory:

	$ hem init

Add a connection profile for *example.com* with manlocal:hem-manage[1] using
your current username (`USER` environment variable), and another connection
profile for *legends.org* with the username *turing*:

	$ hem manage example.com turing@legends.org
	writing ~/.hem/profile/example.com
	authorizing on example.com (~/.ssh/id_dsa.pub)
	Password: *************
	+++ public key added
	writing ~/.hem/profile/turing@legends.org
	authorizing on turing@legends.org (~/.ssh/id_dsa.pub)
	Password: *************
	+++ public key added

Note that, by default, manlocal:hem-manage[1] will attempt to connect to the
remote host and add authorize your public key for future sessions. You may be
prompted for your remote password.

Use manlocal:hem-status[1] to check connection state:

	$ hem status
	example.com                    down
	turing@legends.org             down

Bring up connection to *legends.org* only using manlocal:hem-up[1]:

	$ hem up turing@legends.org
	bringing up: turing@legends.org

Check connection status, again:

	$ hem status
	example.com                    down
	turing@legends.org             up            5432

Bring up all connections:

	$ hem up
	bringing up: example.com
	turing@legends.org is already up

Check connection status, again:

	$ hem status
	example.com                    up            5433
	turing@legends.org             up            5432

Take all connections down with manlocal:hem-down[1]:

	$ hem down
	taking down: example.com (pid: 5433)
	taking down: turing@legends.org (pid: 5432)


ENVIRONMENT VARIABLES
---------------------

`HEM_CONFIG`::
	If `HEM_CONFIG` is set, `hem` uses it as the location of the
	global configuration file. It overrides the default configuration
	file location (*~/.hem/config*), but is overridden by the `-c`
	invocation option.

`HEM_EXEC`::
	The path to hem executables.

FILES
-----

*~/.hem/config*::
	This is the default per user configuration file. The settings in
	this file provide defaults for all connection profiles. Available
	configuration variables are described in the manlocal:hem_config[5]
	manual page.

*~/.hem/profile*::
	The default hem connection profile directory. Each file contains
	configuration for an individual remote SSH connection. Available
	configuration variables are described in the manlocal:hem_profile[5]
	manual page.

*~/.hem/run*::
	Hem stores *.pid* files for running connections in this directory by
	default. The location of this directory can be configured by setting
	the `run_dir` option. See manlocal:hem_config[5] for more information.

*~/.hem/log*::
	Hem logs messages here if no `log_to` option has been specified.

SEE ALSO
--------

manlocal:hem_config[5], man:ssh[1], man:ssh-agent[1], man:ssh_config[5].

AUTHOR
------
Written by http://tomayko.com/[Ryan Tomayko] <r@tomayko.com>

RESOURCES
---------
Web Site
http://tomayko.com/src/hem/[]

Project Page
http://github.com/rtomayko/hem/[]

COPYING
-------
Copyright \(C) 2008, Ryan Tomayko. Free use of this software is
granted under the terms of the New and Simplified BSD License.

// vim: tw=78 ts=8 sts=0 noexpandtab


hem.sh
#!/bin/sh
# hem - ssh controller and automatic tunnel daemon.
#
# Copyright (c) 2008, Ryan Tomayko <r@tomayko.com>
# See the file COPYING included with this package for license information.
set -e

USAGE="[-q][-v][-c <file>] <command> [<opt>]... [<profile>]...
Hem is an ssh controller and tunnel manager.

Available commands:
  init                  initialize the ~/.hem directory
  manage                add a new profile / edit existing profile
  info                  show profile configuration
  status                show connection status

  up                    bring connection profiles up
  down                  take connection profiles down
  bounce                restart connection profiles

Global options:
  -c, --config <file>   load alternative config file
  -q, --quiet           be absolutely quiet
  -v, --verbose         write log messages to stderr
      --version         write program version to stdout and exit
      --help            show usage and exit

See 'hem <command> --help' for information on a specific <command>"

# initialize HEM_DIR and HEM_CONFIG to their default values.
HEM_EXEC=${HEM_EXEC:-"@@HEM_EXEC_DIR@@"}
HEM_VERSION="@@HEM_VERSION@@"
PATH="$HEM_EXEC:$PATH"

. hem-sh-setup

# the default config file location.
HEM_CONFIG=${HEM_CONFIG:-"$HOME/.hem/config"}

# when set, info messages are discarded.
quiet=

# when set, log messages go to stderr along with the log.
verbose=

# commands that do not require valid profile names as arguments.
simple_commands="init list manage help"

# with no arguments, bail out with usage
if [ $# -eq 0 ] ; then
	usage
	exit 1
fi

# parse top-level arguments...
while [ $# -gt 0 ]
do
case "$1" in
	-c|--config)
		HEM_CONFIG="$2"
		shift; shift
		;;
	-q|--quiet)
		quiet=1
		shift
		;;
	-v|--verbose)
		quiet=1
		shift
		;;
	--version)
		echo "Hem $HEM_VERSION"
		exit 0
		;;
	-h|--h*)
		usage
		exit 0
		;;
	-*)
		see_usage "invalid argument: $1"
		;;
	*)
		break
		;;
esac
done

# determine command
command=
if [ $# -gt 0 ]; then
	command="$1"
	shift
else
	see_usage "command not specified."
fi

# verify that command exists
command_path="$HEM_EXEC/hem-$command"
[ -x "$command_path" ] ||
see_usage "unknown command: $command"

[ -n "$(echo "$simple_commands" | grep $command 2>/dev/null)" ] &&
simple_command=1 ||
simple_command=

# parse command arguments
command_argv=
command_help=
profiles=
while [ $# -gt 0 ]; do
case "$1" in
	-h|--hel*)
		command_help=1
		command_argv="$command_argv $1"
		shift
		;;
	--)
		shift
		break
		;;
	-*)
		command_argv="$command_argv $1"
		shift
		;;
	*)
		break
esac
done

# all remaining arguments are profile names. command will be run once
# per profile.
profiles="$@"

# ---------------------------------------------------------------------
# Global Config
# ---------------------------------------------------------------------
#
# Hem's default config file location is ~/.hem/config but it can be
# overridden on the command line or by setting HEM_CONFIG. Hem does not
# require a config file so care must be taken to not assume one
# exists and that defaults are provided.

test -r "$HEM_CONFIG" && {
	__FILE__="$HEM_CONFIG"
	. "$HEM_CONFIG"
	unset __FILE__
}

# configure default config values
hem_dir=$(dirname $HEM_CONFIG)
log_to=${log_to:-}
run_dir=${run_dir:-"$hem_dir/run"}
state_dir=${state_dir:-"$hem_dir/state"}
profile_dir=${profile_dir:-"$hem_dir/profile"}
poll_time=${poll_time:-600}

# setup sub-command environment
export HEM_CONFIG PATH quiet verbose
export log_to run_dir state_dir profile_dir poll_time

# If command help is requested, exec the command immediately.
if [ $command_help ]; then
	exec $command_path --help
fi

# if this is a simple command, invoke it immediately without
# gathering profiles or verifying anything else.
if [ $simple_command ]; then
	exec $command_path $command_argv $profiles
fi

# figure out what profiles we're operating on..
if test -z "$profiles" ||
   test "$profiles" = "--all" ||
   test "$profiles" = "all"
then
	test -d "$profile_dir" &&
	profiles="$( cd "$profile_dir" && ls -1 | grep -v '~$' || true )"
fi

# Loop over selected profiles, kicking off command for each.
failures=0
result=
set +e
for profile in $profiles ;
do
	"$command_path" $command_argv "$profile"
	result=$?
	test "$result" = 1 &&
	failures=$(expr $failures + 1)
done

# exit based on whether any of the commands failed
if [ $failures -gt 1 ] ; then
	die "multiple failures."
elif [ $failures -eq 1 ] ; then
	exit 1
else
	exit 0
fi
